Large misc refactor

Jonxslays opened the pull request:
This PR refactors the code base to bring it into compliance with [PEP 8](https://peps.python.org/pep-0008/).

- Added strict type checking, and moved the project into a new `hom`module.
- Added readme, updated dependencies.
- Added nox for linting and type checking.
- Added github CI for linting and type checking.
- Added example environment file.
- Added command for sending the initial support embed.
- Added README.

I tested the functionality pretty thoroughly, seems to all be working still.

I would like to add a license before this PR gets merged, what do we want to license this under @psikoi @Aluminoti?

Jonxslays committed:
Update requirements file

Jonxslays committed:
Add example .env file

Jonxslays committed:
Update example env

Jonxslays committed:
Update requirements files

Jonxslays committed:
Add comments to example env

Jonxslays committed:
Add readme

Jonxslays committed:
Add empty noxfile

Jonxslays committed:
Add pyproject.toml

Jonxslays committed:
Move wom support and views into hom module

Jonxslays committed:
Remove data module

Jonxslays committed:
Partial rewrite of utilities

Jonxslays committed:
Add noxfile linting

Jonxslays committed:
Migrate utilites and update views

Jonxslays committed:
Update dependencies

Jonxslays committed:
Add sync method to bot

Jonxslays committed:
Typing fixes on views

Jonxslays committed:
Load dotenv in config file

Jonxslays committed:
Update noxfile

Jonxslays committed:
Utils typing fixes

Jonxslays committed:
Slim down module entry point

Jonxslays committed:
Port old data cog to new support cog

Jonxslays committed:
Remove unused guild id from environment and config

Jonxslays committed:
Update readme

Jonxslays committed:
Add linting workflow

Jonxslays committed:
Remove debug prints

Jonxslays committed:
Add command error handling

Jonxslays committed:
Update branch name

Jonxslays committed:
Add command for sending support message

Jonxslays committed:
Rename sticky channel to support channel

Jonxslays committed:
Add dunder all

Jonxslays committed:
Add editorconfig

Jonxslays committed:
Fix windows bug with workflows

Jonxslays committed:
Update uvloop comment

Jonxslays committed:
Rename example env for syntax highlighting

Jonxslays committed:
Update readme

Jonxslays committed:
Fix bug with mog log messages

Jonxslays committed:
Update error message

rorro commented on README.md:
````diff
@@ -0,0 +1,36 @@
+
+```bash
+$ python -m hom
+```
````
After running the bot, one must run `!sync` to register the slash commands. I think it should be included in the readme. Else you'd have to read the source code to find this out.

Jonxslays replied:
Should we just sync the commands on bot startup automatically without the need for the command?

rorro commented on hom/utils.py:
```diff
@@ -0,0 +1,267 @@
+    tickets_category = get_category(interaction.guild, "Tickets")  # TODO: Make this channel_id?
+    mod_role = get_role_by_name(interaction.guild, "Moderator")
+    helpful_old_man_role = get_role_by_name(interaction.guild, "Helpful Old Man")
+    assert helpful_old_man_role
+    assert mod_role
```
These asserts will fail unless I have my roles set to `Moderator` and `Helpful Old Man`. Maybe it would be better to use role IDs instead and define them in `.env`.

Jonxslays replied:
yes, role ids will be better but the asserts are still necessary. because those roles need to be there for the bot to function.

rorro commented on hom/utils.py:
```diff
@@ -0,0 +1,267 @@
+    channel: t.Optional[discord.TextChannel] = None,
+) -> t.Optional[discord.Message]:
+    assert interaction.guild
+    mod_logs_channel = get_channel_by_name(interaction.client, interaction.guild.id, "mod-logs")
```
Same as with the roles, wouldn't it be better to define the channel id in `.env` instead of forcing the name of the channel to `mod-logs`? If the channel name ever changes for any reason, this feature will break.

Jonxslays commented:
all valid comments @rorro there's a TODO comment from aluminoti somewhere talking about using role id's too.

Aluminoti commented:
> I would like to add a license before this PR gets merged, what do we want to license this under @psikoi @Aluminoti?

I'm not very familiar with any licensing things. I'm more/less open to whatever @psikoi or others suggest.

Jonxslays commented:
I typically just go with MIT, BSD 3 Clause, or GPLv3

Jonxslays committed:
Add MIT license

Jonxslays committed:
Update readme

Jonxslays committed:
Add additional config env vars

Jonxslays committed:
Add questions channel to env

Jonxslays committed:
Add Forbidden to handled errors

Jonxslays committed:
Refactor utils

Jonxslays committed:
Update cogs to use new utils functions

Jonxslays commented:
- Added `!sync` command into the readme.
- Refactored to use role and channel id's in the environment.
- Added MIT License with copyright holder as @psikoi. If you want that changed just let me know.

Aluminoti reviewed (commented):
This is probably fine but not what the bot is currently using. I haven't looked in to any changes between either of these versions.

Aluminoti commented on hom/utils.py:
```diff
@@ -0,0 +1,181 @@
+    channel_name = channel_name.lower().replace("-", " ")
+    return discord.utils.find(
+        lambda c: str(c.name).lower().replace("-", " ") == channel_name,
+        guild.categories,  # JONX_TODO: Changed this to categories, is that right?
```
This is correct, yes.

Jonxslays replied:
Somehow you're reviewing an old version of these changes, that comment doesn't exist in the PR anymore. I completely rewrote that function.

Aluminoti commented on hom/cogs/views.py:
```diff
@@ -14,10 +15,10 @@ def __init__(self) -> None:
     )
     async def groups_instructions(
-        self, interaction: discord.Interaction, button: discord.ui.Button
+        self, interaction: discord.Interaction, _: discord.ui.Button["Support"]
```
Can this be explained to me? Not quite sure what the ["Support"] does.

Jonxslays replied:
`Button` is generic over the View it is implemented on. In this case the `Support` View.

Aluminoti commented on hom/cogs/views.py:
```diff
@@ -208,33 +252,41 @@ def __init__(self):
+        embed = discord.Embed(description=f"{interaction.user.mention} has closed the ticket.")
+
+        assert isinstance(interaction.channel, discord.guild.GuildChannel)
+        assert isinstance(interaction.user, discord.Member)
```
What do these assert statements do if they fail? Not really familiar with the purpose of assert statements so I'm not sure if it's valid but my main concern here would be never responding to the interaction.

Assuming the asserts fail something, the asserts could probably be moved to the top, pushing the embed creation below them.

Jonxslays replied:
assert throws an `AssertionError` if whatever is asserted is not truthy.

Jonxslays replied:
Moved them to the top.

Aluminoti commented on hom/utils.py:
```diff
@@ -48,63 +53,73 @@ def get_channel_by_id(
+    if not any(guild_id == guild.id for guild in bot.guilds):
+        return None
+
+    if not (guild := bot.get_guild(guild_id)):
```
Does this assign guild and then check that guild is "something" and not "nothing"? If so, very cool. If not, the "guild.channels" below is lacking context.

Jonxslays replied:
Yes that's exactly what it does.

Aluminoti commented on hom/utils.py:
```diff
@@ -48,63 +53,73 @@ def get_channel_by_id(
         category=tickets_category,
         reason=f"{interaction.user.display_name} ({interaction.user.id}) has opened a ticket.",
-        topic=button_label,
+        topic=button_label or "Unknown (This is a bug)",
```
Something tells me this isn't entirely accurate. If we provide an empty string for the button_label being provided (not sure if we would/should/do) then that's not necessarily "Unknown (This is a bug)", it would just be a blank channel topic.

I think the more correct thing to do would be to require a topic parameter when using the provided method.

Jonxslays replied:
There should always be a button label. that's why if there isn't it's a bug.

In our case all our buttons use a label, but label isn't required and discord.py types them as `Optional` meaning to be type safe we need to handle the case where it could be `None` hence the `or "Unknown (this is a bug)"`.

Aluminoti commented on hom/utils.py:
```diff
@@ -119,58 +134,71 @@ async def create_ticket_for_user(interaction, instructions, button_label, exampl
+    mod: discord.User | discord.ClientUser | discord.Member,
+    channel: t.Optional[discord.guild.GuildChannel] = None,
+) -> t.Optional[discord.Message]:
+    assert interaction.guild
```
Can an interaction ever not have a guild? (again, still unfamiliar with assert statements)

Jonxslays replied:
If it's used outside of a guild yes - which we don't want. All our commands should always be run inside the guild and if they're not we bail out.

Aluminoti commented on hom/utils.py:
```diff
@@ -40,37 +40,41 @@ def get_category(guild: discord.Guild, channel_name: str) -> t.Optional[discord.
     channel_name: str,
     channel_type: str = "text",
-) -> t.Optional[discord.guild.GuildChannel]:
+) -> t.Optional[discord.TextChannel]:
```
This isn't accurate if we ever use a different "channel_type". Might be addressed in a later commit but it seems like maybe the more correction option for us to do would be to remove the channel_type and strictly use this for getting text channels and get_category() for categories.

All that said, not sure if "-> t.Optional[discord.TextChannel]:" literally means it may not be a text channel and that we are defaulting the type to a text channel because the channel_type parameter is defaulted to "text".

Jonxslays replied:
This function no longer exists also. But if the ID in the environment is wrong we will never find the channel. That's why its optional.

Aluminoti commented on hom/utils.py:
```diff
@@ -9,6 +9,21 @@
 from hom.cogs import views
 from hom.config import Constants
 
+__all__ = (
```
Can I get a brief explanation as to what this is used for? I might answer my own question when reviewing everything all at once but sometimes an explanation can be better than assumption.

Jonxslays replied:
if you we're to `from utils import *` everything in the `__all__` is what you what get imported.

Aluminoti commented on hom/utils.py:
```diff
@@ -120,37 +68,37 @@ async def create_ticket_for_user(
         topic=button_label or "Unknown (This is a bug)",
         overwrites={
+            mod_role: discord.PermissionOverwrite(read_messages=True),
+            interaction.guild.me: discord.PermissionOverwrite(read_messages=True),
```
Not sure if the change here is just going from interaction.user to interaction.guild.me - I'm sure both work but would be curious if there is a specific reason it was changed.

Jonxslays replied:
No, originally we we're fetching the bots role by name to give permissions to the channel to that role. It was a bit clunky, instead we just give them bot itself access to the channel based on it's user. 

Additionally, you can't pass a `User` to overwrites, only a `Member` or `Role`, which is why we gets the bots `Member` (interaction.guild.me).

Aluminoti commented on hom/cogs/support.py:
```diff
@@ -76,13 +76,15 @@ async def clear_ratelimit(self, user_id: int) -> None:
         assert isinstance(interaction.user, discord.Member)
 
-        if not (is_mod := utils.contains_roles(interaction.user.roles, "Moderator")):
+        if not any(r.id == Config.MOD_ROLE for r in interaction.user.roles):
```
Might be a good idea to update Config.MOD_ROLE to Config.MOD_ROLE_ID since it's not really a role.

Jonxslays replied:
Everything pulled from the environment is an ID, except the token. I don't really think it's a big deal.

Jonxslays commented:
> This is probably fine but not what the bot is currently using. I haven't looked in to any changes between either of these versions.

If you're referring to the bump in discord.py version, yes it's fine. Otherwise, I'm unsure what you're referring to.

Jonxslays committed:
Move asserts to top of methoc

Jonxslays committed:
Remove Bot.sync method that was only used once

rorro commented on hom/cogs/views.py:
```diff
@@ -0,0 +1,344 @@
+            "Some name changes get skipped, as they can’t be auto-approved by our system and "
+            "require manual approval.\n\nIf yours hasn’t been auto-approved, please tell us the "
+            "name change ID and we’ll manually review it for you.\n\nNote: If you’d like to know "
+            "why your name change has been skipped you can visit our beta website (work in "
+            "progress) at https://beta.wiseoldman.net/names and hover your cursor over the your "
```
The new site has been released and this text should be changed a little.

Jonxslays replied:
good catch!

rorro reviewed (changes requested):
Other than the small text change in view.py, it looks good to me. Also tested it and everything works as expected.

Jonxslays committed:
Remove reference to beta from name change embed

Jonxslays committed:
Use mod role id instead of name for sync command
